fix(networking): don't report ConnectionIssues during initial bootstrap

joshuef opened the pull request:
context from the forums: https://forum.autonomi.community/t/basiceconomytweaks-early-technical-beta/39528/382?u=joshuef

## Description

reviewpad:summary

grumbach commented on sn_node/src/node/mod.rs:
```diff
@@ -554,6 +553,32 @@ mod core {
 
             let mut cmds = vec![];
 
+            // clean up DKG sessions 2 generations older than current
+            // `session_id.section_chain_len + 2 < current_chain_len`
+            // we voluntarily keep the previous DKG rounds
+            // so lagging elder candidates can still get responses to their gossip.
+            // At generation+2, they are not going to be elders anymore so we can safely discard it
+            let current_chain_len = self.network_knowledge.section_chain_len();
+            let mut old_hashes = vec![];
+            for (hash, session_info) in self.dkg_sessions_info.iter() {
+                if session_info.session_id.section_chain_len + 5 < current_chain_len {
```
Ideally we should only need +2, but for now, before we fix the AE issue after selfish terminations +5 is a safer bet.

davidrusu commented on sn_interface/src/messaging/system/agreement.rs:
```diff
@@ -66,6 +65,11 @@ impl DkgSessionId {
         hash
     }
 
+    /// A small checksum of the session id's hash used for logging as it is very short
+    pub fn sum(&self) -> u16 {
+        self.hash().iter().map(|i| *i as u16).sum::<u16>()
+    }
+
```
Give this a more descriptive name, something like `dbg_short_hash`. `sum` makes it seem like something general / important.

Also, you'll get fewer collisions if you just take the prefix of the hash, i.e. if you want 16 bits, take the first two bytes of the hash.
```suggestion
    /// A small checksum of the session id's hash used for logging as it is very short
    pub fn dbg_short_hash(&self) -> u16 {
        u16::from_le_bytes(self.hash()[..2])
    }

```

grumbach replied:
or maybe `sh()`

davidrusu commented on sn_interface/src/network_knowledge/mod.rs:
```diff
@@ -94,6 +94,12 @@ pub const fn supermajority(group_size: usize) -> usize {
 
+/// `Threshold` of a given group
+#[inline]
+pub const fn threshold(group_size: usize) -> usize {
+    supermajority(group_size) - 1
+}
```
I don't like seeing subtraction with `usize`, Let's define `supermajority` in terms of `threshold` to get rid of it.

Currently `supermajority = 1 + group_size * 2 / 3` and `threshold = supermajority - 1`
Instead do `supermajority = threshold + 1` and `threshold = group_size * 2 / 3`

davidrusu commented on sn_node/src/node/dkg.rs:
```diff
@@ -0,0 +1,362 @@
+    session_id: &DkgSessionId,
+    participant_index: usize,
+    secret_key: BlsSecretKey,
+    pub_keys: DkgPubKeys,
```
There's a few things named some variation of `public-keys` in this function and it's kind of hard to track. Can we rename some by functionality?

```suggestion
    node_key_upgrades: DkgPubKeys,
```

davidrusu commented on sn_node/src/node/dkg.rs:
```diff
@@ -0,0 +1,362 @@
+    let mut rng = bls::rand::rngs::OsRng;
+    let threshold = threshold(session_id.elders.len());
+    let mut public_keys: BTreeMap<NodeId, BlsPublicKey> = BTreeMap::new();
+    for (xorname, (pubkey, _)) in pub_keys.iter() {
```
```suggestion
    for (xorname, (ephemeral_bls_pk, _)) in pub_keys.iter() {
```

davidrusu commented on sn_node/src/node/dkg.rs:
```diff
@@ -0,0 +1,362 @@
+}
+
+// Helper that checks an ephemeral pubkey
+pub(crate) fn check_key(
```
```suggestion
pub(crate) fn check_ephemeral_dkg_key(
```

davidrusu commented on sn_node/src/node/dkg.rs:
```diff
@@ -0,0 +1,362 @@
+        session_id.sum(),
+        key_owner
+    );
+    let serialized = bincode::serialize(&key)?;
```
```suggestion
    let serialized_key = bincode::serialize(&key)?;
```

davidrusu commented on sn_node/src/node/dkg.rs:
```diff
@@ -0,0 +1,362 @@
+        // gen new key
+        let new_secret_key: BlsSecretKey = bls::rand::random();
+        let new_pub_key = new_secret_key.public_key();
+        let serialized = bincode::serialize(&new_pub_key)?;
```
```suggestion
        let serialized_key = bincode::serialize(&new_pub_key)?;
```

davidrusu commented on sn_node/src/node/dkg.rs:
```diff
@@ -0,0 +1,362 @@
+        let new_secret_key: BlsSecretKey = bls::rand::random();
+        let new_pub_key = new_secret_key.public_key();
+        let serialized = bincode::serialize(&new_pub_key)?;
+        let sig = types::keys::ed25519::sign(&serialized, keypair);
```
```suggestion
        let key_sig = types::keys::ed25519::sign(&serialized, keypair);
```

davidrusu commented on sn_node/src/node/dkg.rs:
```diff
@@ -0,0 +1,362 @@
+
+pub(crate) type DkgPubKeys = BTreeMap<XorName, (BlsPublicKey, Signature)>;
+
+pub(crate) struct DkgEphemeralKeys {
+    secret_key: BlsSecretKey,
+    pub_keys: DkgPubKeys,
+}
```
Why do we store the `pub_keys` as a map? I'd imagine this struct would be:
```suggestion
pub(crate) struct DkgEphemeralKeys {
    owner: XorName,
    secret_key: BlsSecretKey,
    signature: Signature
}
```
You don't need to hold the public key since it can be re-computed from the secret_key with `secret_key.public_key()`

davidrusu commented on sn_node/src/node/dkg.rs:
```diff
@@ -0,0 +1,362 @@
+            pub_keys: BTreeMap::from_iter([(our_name, (new_pub_key, sig))]),
+        };
+
+        // insert the key if we don't already have it
+        let _did_insert = self
+            .dkg_ephemeral_keys
+            .insert(session_id_hash, ephemeral_keys);
```
Is the comment outdated? This code always inserts the new ephemeral key.

grumbach replied:
That comment line is probably not very useful, let me remove it. It's redundant with the bail out above if we already have an entry for this hash.

davidrusu commented on sn_node/src/node/messaging/dkg.rs:
```diff
@@ -107,176 +150,567 @@ impl Node {
+        );
+        let peers = dkg_peers(our_id, session_id);
+        let node_msg = SystemMsg::DkgVotes {
+            session_id: session_id.clone(),
+            pub_keys,
+            votes: vec![vote],
         };
-        self.send_system_msg(msg, Peers::Single(sender))
+        let cmd = self.send_system_msg(node_msg, Peers::Multiple(peers));
+        Ok(vec![cmd])
```
Does this initial vote also get processed by us at some point?

davidrusu commented on sn_node/src/node/messaging/dkg.rs:
```diff
@@ -107,176 +150,567 @@ impl Node {
+            session_id.sh()
+        );
+        let peers = dkg_peers(our_id, session_id);
+        let node_msg = SystemMsg::DkgVotes {
+            session_id: session_id.clone(),
+            pub_keys,
+            votes: vec![vote],
         };
-        self.send_system_msg(msg, Peers::Single(sender))
+        let cmd = self.send_system_msg(node_msg, Peers::Multiple(peers));
```
Can this be a call to `broadcast_dkg_votes(..)`?

grumbach replied:
Oh good one!

davidrusu reviewed (approved):
nice work :1st_place_medal:

davidrusu commented:
bors r+

davidrusu commented:
@grumbach you've included a very large `save.svg` statemap! lets remove that before merging.

bors cancel

grumbach commented:
bors r+

davidrusu commented:
bors r+

joshuef committed:
fix(networking): don't report ConnectionIssues during initial bootstrap

context from the forums:
https://forum.autonomi.community/t/basiceconomytweaks-early-technical-beta/39528/382
